### PR TITLE
Tests: Improve PHPUnit Forward Compatibility

### PR DESCRIPTION
**Pull request by swissspidy**
**PHPUnit 8.x and 9.x compatibility**

tl;dr: we need to update our PHP integration tests running against core so that they function properly (again) as needed.

WordPress core has moved to adopt PHPUnit polyfills for running its tests and at the same time runs tests using the latest possible PHPUnit version.

To make our test suite compatible, we need to:

* Update assertions so we don't use deprecated ones
* Define `WP_TESTS_PHPUNIT_POLYFILLS_PATH`
* Update PHPUnit to latest possible version on CI
* Update PHPUnit configuration to improve forward compatibility
* Update `Yoast/wp-test-utils` package

The PR also includes one small change related to PHP 8.1 compatibility, see #8520.

See https://core.trac.wordpress.org/ticket/46149 and some of the early individual commit messages for details.

See also https://core.trac.wordpress.org/ticket/53911 for when the PHPUnit changes were backported to older branches

### CODE REVIEW

File: includes/Renderer/Stories/Renderer.php [hunk 1]
```diff
@@ -37,6 +37,9 @@
 use Iterator;
 use WP_Post;
 
+// Disable reason: prevent false positives due to ReturnTypeWillChange attributes.
```
Comment (spacedmonkey) [inline includes/Renderer/Stories/Renderer.php:40]: Can you explain this one a little better?

File: phpunit-integration-multisite.xml.dist [hunk 2]
```diff
@@ -1,11 +1,14 @@
 <?xml version="1.0" ?>
 <phpunit
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
   bootstrap="tests/phpunit/integration/bootstrap.php"
   backupGlobals="false"
   colors="true"
+  failOnRisky="true"
   convertErrorsToExceptions="true"
   convertNoticesToExceptions="true"
   convertWarningsToExceptions="true"
```
Comment (spacedmonkey) [inline phpunit-integration-multisite.xml.dist:10]: Worth doing [this](https://twitter.com/jrf_nl/status/1441823275550527494) at the same time.

File: tests/phpunit/integration/tests/Experiments.php [hunk 3]
```diff
@@ -39,10 +39,6 @@ public static function wpSetUpBeforeClass( $factory ) {
 		);
 	}
 
-	public static function wpTearDownAfterClass() {
-		self::delete_user( self::$user_id );
-	}
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/Experiments.php:44]: Why is this removed?

File: tests/phpunit/integration/tests/Media/SVG.php [hunk 4]
```diff
@@ -23,44 +23,58 @@
...
+	protected $instance;
+
+	public function set_up() {
+		parent::set_up();
+
+		$experiments = $this->createMock( \Google\Web_Stories\Experiments::class );
+		$experiments->method( 'is_experiment_enabled' )
+					->willReturn( true );
+
+		$this->instance = new \Google\Web_Stories\Media\SVG( $experiments );
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/Media/SVG.php:39]: I don't like this. It is much cleaner to create a new instance of a object every time for each test. So there no problem of state being from previous test. The test was better the way it was.

File: tests/phpunit/integration/tests/REST_API/Stories_Controller.php [hunk 5]
```diff
@@ -87,13 +88,11 @@ public static function wpSetUpBeforeClass( $factory ) {
...
 		);
 
-		$future_date = strtotime( '+1 day' );
-
 		$factory->post->create_many(
 			3,
 			[
 				'post_status' => 'future',
-				'post_date'   => strftime( '%Y-%m-%d %H:%M:%S', $future_date ),
+				'post_date'   => ( new DateTime( '+1day' ) )->format( 'Y-m-d H:i:s' ),
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/REST_API/Stories_Controller.php:95]: Why is this being changed?

File: tests/phpunit/integration/tests/Story_Post_Type.php [hunk 6]
```diff
@@ -66,79 +73,77 @@ public static function wpSetUpBeforeClass( $factory ) {
 		set_post_thumbnail( self::$story_id, $poster_attachment_id );
 	}
 
-	public function setUp() {
-		parent::setUp();
+	public function set_up() {
+		parent::set_up();
+
+		$this->instance = new \Google\Web_Stories\Story_Post_Type();
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/Story_Post_Type.php:91]: Again, it is a bad idea to reuse objects like this. New test new instance.

**Comment by spacedmonkey** [comment]
This PR needs to be rebased. 

There seems to be lots of places where you have converted to use a single instance of the class. This bad for unit testing and should be reverted. What was the reasoning behind this change in the first place?

Also, all the instance where users can deleted after tests are deleted, why were these removed?

File: includes/Renderer/Stories/Renderer.php (hunk 1 above)
Comment (swissspidy) [reply includes/Renderer/Stories/Renderer.php:40]: Because there now is `#[\ReturnTypeWillChange]` between the docblocks and the actual methods, PHPCS wrongly thinks the methods don't have any docblocks associated with them.

`#[\ReturnTypeWillChange]` annotations are needed for PHP 8.1 compatibility. See https://wiki.php.net/rfc/internal_method_return_types

File: phpunit-integration-multisite.xml.dist (hunk 2 above)
Comment (swissspidy) [reply phpunit-integration-multisite.xml.dist:10]: This change here actually addresses the issue described in that tweet. See the linked core ticket, https://core.trac.wordpress.org/ticket/54183.

This change here is option "fail on risky tests caused by deprecation warnings"

So we're all set there

File: tests/phpunit/integration/tests/Experiments.php (hunk 3 above)
Comment (swissspidy) [reply tests/phpunit/integration/tests/Experiments.php:44]: WordPress already deletes all users in `\WP_UnitTestCase_Base::tear_down_after_class`

So this is redundant.

File: tests/phpunit/integration/tests/REST_API/Stories_Controller.php (hunk 5 above)
Comment (swissspidy) [reply tests/phpunit/integration/tests/REST_API/Stories_Controller.php:95]: `strftime` is deprecated in PHP 8.1

File: tests/phpunit/integration/tests/Story_Post_Type.php (hunk 6 above)
Comment (swissspidy) [reply tests/phpunit/integration/tests/Story_Post_Type.php:91]: Keeping dicussion in one place at https://github.com/google/web-stories-wp/pull/8607#discussion_r716849381

File: tests/phpunit/integration/tests/Media/SVG.php (hunk 4 above)
Comment (swissspidy) [reply tests/phpunit/integration/tests/Media/SVG.php:39]: I was doing some cleanup like this while working on this PR to make things more DRY.

Note that this is in `set_up` so we still get a new instance for every test. No need to repeat this when `set_up` can do this once for every test.

So I disagree with the "The test was better the way it was"

File: .github/workflows/tests-unit-php.yml [hunk 7]
```diff
@@ -81,10 +90,10 @@ jobs:
       # So we temporarily switch PHP versions, do a full install and then remove the package.
       # Then switch back to the PHP version we want to test and delete the vendor directory.
 
-      - name: Setup PHP 7.4
+      - name: Setup PHP 8.0
         uses: shivammathur/setup-php@v2
         with:
-          php-version: 7.4
+          php-version: '8.0'
```
Comment (spacedmonkey) [inline .github/workflows/tests-unit-php.yml:96]: Does this need quote marks?

File: phpunit-integration-multisite.xml.dist (hunk 2 above)
Comment (spacedmonkey) [reply phpunit-integration-multisite.xml.dist:10]: Good.

**Review by spacedmonkey — APPROVED** [review]
This looks great!
